### PR TITLE
CRM Automations: Bootstrap system

### PR DESCRIPTION
**Pull request by kallehauge**
Fixes https://github.com/Automattic/zero-bs-crm/issues/3263

## Proposed changes:

* Add a new `Bootstrap` class that is solely responsible for:
  * Initialise the automation engine
  * Register "core triggers/conditions/actions" and allow 3. parties to hook into it 
* Remove the need to inject `Automation_Engine` into workflow constructs
  * The intent here is to make it a smoother experience programmatically registering workflows.
  * This also means that we had to add new set/get logger methods to a workflow
* Update existing workflow tests to inject the engine via `Automation_Workflow::set_engine`
* Add a new test that ensures that we throw a fatal error if someone tries to execute a workflow without injecting the engine.
* Add a `is_admin` condition around the automation bootstrap logic so we do not bootstrap all of it on front-end page loads.
  * I've added a `@todo` condition since I believe this could be a general issue for CRM.

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

pbhBOz-2ES-p2

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

* Add the below snippet somewhere in your code (e.g. `0-sandbox.php).
* _(Optional) Enable error log and tail it to verify that we do not throw any errors._
* Go to `/wp-admin/admin.php?page=zbs-add-edit&action=edit&zbstype=contact and add a new contact
  * The status should be anything else than `Customer` to verify the workflow worked.
* Verify that the status was set to `Customer` by the workflow.

```php
// Enable the automation system.
add_filter( 'jetpack_crm_feature_flag_automations', '__return_true' );

// Programmatically define a workflow that will always set a contacts status to "Customer" when they are first created.
add_filter(
	'jpcrm_automation_workflows',
	function ( $workflows ) {
		$workflow = new \Automattic\Jetpack\CRM\Automation\Automation_Workflow(
			[
				'id'           => 'testing',
				'name'         => __( 'New contact: Set status to "Customer"', 'zero-bs-crm' ),
				'description'  => __( 'This automation will change the status of a contact to "Customer" when they are created.', 'zero-bs-crm' ),
				'category'     => 'contact',
				'triggers'     => [
					\Automattic\Jetpack\CRM\Automation\Triggers\Contact_Created::get_slug(),
				],
				'initial_step' => [
					'slug'            => 'contact_status_condition',
					'class_name'      => \Automattic\Jetpack\CRM\Automation\Actions\Update_Contact_Status::class,
					'attributes'      => [
						'new_status' => 'Customer',
					],
					'next_step_true'  => null,
					'next_step_false' => null,
				],
			]
		);

		$workflows[ $workflow->get_id() ] = $workflow;

		return $workflows;
	}
);
```

### CODE REVIEW

**Comment by cleacos** [comment]
We will need these changes for our current working code. Before merging this one, it'd like to finish the review of the [Quote triggers PR](https://github.com/Automattic/jetpack/pull/32099), and merge it.

File: projects/plugins/crm/src/automation/class-automation-boostrap.php
Comment (cleacos) [inline projects/plugins/crm/src/automation/class-automation-boostrap.php:125]: Future note: The step register will be the same method used for registering triggers, without the need of calling `get_slug`. Only one parameter.

File: projects/plugins/crm/src/automation/class-automation-workflow.php
Comment (cleacos) [inline projects/plugins/crm/src/automation/class-automation-workflow.php:15]: This should be int... is the row ID in the database.

File: projects/plugins/crm/src/automation/class-automation-workflow.php
Comment (cleacos) [inline projects/plugins/crm/src/automation/class-automation-workflow.php:119]: OK. The id was added for the database row. Normally I try to follow the strict type way and avoid mixed types. Do you have a use case where we need the ID for hardcoded workflows? `$name` could be enough? If so, any format for it?

File: projects/plugins/crm/src/automation/class-automation-workflow.php
Comment (cleacos) [inline projects/plugins/crm/src/automation/class-automation-workflow.php]: Should we translate the technical log?

Comment (cleacos) [reply projects/plugins/crm/src/automation/class-automation-workflow.php]: If the output is for the end user, I'd say yes, absolutely, but if it is for debugging purposes I'd keep it in English.

**Comment by cleacos** [comment]
I've updated some tests that were still using the Automation_Workflow constructor with the Automation_Engine parameter.

File: projects/plugins/crm/src/automation/class-automation-boostrap.php
Comment (kallehauge) [reply projects/plugins/crm/src/automation/class-automation-boostrap.php:125]: 👍  I noticed it as well, but the PR seemed to get out of scope if I corrected it here 😄

Comment (cleacos) [reply projects/plugins/crm/src/automation/class-automation-boostrap.php:125]: Yes, sure! It's not for this PR.

Comment (cleacos) [reply projects/plugins/crm/src/automation/class-automation-boostrap.php:125]: Issue added to the project board

**Comment by cleacos** [comment]
I've updated the testing code to add in `0-sandbox.php` file:

```
add_filter(
	'jpcrm_automation_workflows',
	function ( $workflows ) {
		$workflow = new \Automattic\Jetpack\CRM\Automation\Automation_Workflow(
			[
				'id'           => 'testing',
				'name'         => __( 'New contact: Set status to "Customer"', 'zero-bs-crm' ),
				'description'  => __( 'This automation will change the status of a contact to "Customer" when they are created.', 'zero-bs-crm' ),
				'category'     => 'contact',
				'triggers'     => [
					`jpcrm/contact_created`,
				],
				'initial_step' => [
					'class_name'      => \Automattic\Jetpack\CRM\Automation\Actions\Update_Contact_Status::class,
					'attributes'      => [
						'new_status' => 'Customer',
					],
					'next_step'  => null,
				],
			]
		);

		$workflows[ $workflow->get_id() ] = $workflow;

		return $workflows;
	}
);
```

Currently, it doesn't work because of the hook name change in the Contact_Created trigger, the listen_to_event should listen to `jpcrm_automation_contact_new`, but it's going to be changed by the Event Manager PR. Changing it manually the test works as expected, on Contact creation it is set as Customer.

**Note:** For the steps, we can use the slug name or the class name, both work independently. The class_name has preferent over the slug when the step is executed, it'll try to instance the class name passed, and the fallback is the slug name.

**Comment by cleacos** [comment]
Other than that, the code looks good and works as expected. Let's merge it ASAP to be able to continue making progress with real use cases.

Let's wait for the Event Manager code to be merged in trunk, and then update this PR.

File: projects/plugins/crm/src/automation/class-automation-workflow.php
Comment (kallehauge) [reply projects/plugins/crm/src/automation/class-automation-workflow.php:119]: Honestly, then it's still a bit TBD depending on how we end up storing attributes for "internal workflows". E.g. if we ever added something like the example below.

```php
add_filter(
	'jpcrm_automation_workflows',
	function ( $workflows ) {
		// Contact Created: Force contact status to be "New".
		$workflow = new \Automattic\Jetpack\CRM\Automation\Automation_Workflow(
			[
				'id'           => 'testing',
				'name'         => __( 'New contact: Set status to "Customer"', 'zero-bs-crm' ),
				'description'  => __( 'This automation will change the status of a contact to "Customer" when they are created.', 'zero-bs-crm' ),
				'category'     => 'contact',
				'triggers'     => [
					\Automattic\Jetpack\CRM\Automation\Triggers\Contact_Created::get_slug(),
				],
				'initial_step' => [
					'slug'            => 'contact_status_condition',
					'class_name'      => \Automattic\Jetpack\CRM\Automation\Actions\Update_Contact_Status::class,
					'attributes'      => [
						'new_status' => 'Customer',
					],
					'next_step_true'  => null,
					'next_step_false' => null,
				],
			]
		);

		$workflows[ $workflow->get_id() ] = $workflow;

		return $workflows;
	}
);
```

In this case, we'd have to rely on the manually defined ID to identify which workflow we're potentially updating attributes for via our APIs - and we cannot make it a number since it doesn't actually exist in the database.

We could end up storing everything in the database and in that case we'd be able to be strict about

File: projects/plugins/crm/src/automation/class-automation-workflow.php
Comment (kallehauge) [reply projects/plugins/crm/src/automation/class-automation-workflow.php]: Summary from 1:1 conversation: We're sticking with "machine friendly" messages (read: we won't translate them).
The UI that handles automations should provide guidance instead.

File: projects/plugins/crm/src/automation/class-automation-workflow.php
Comment (cleacos) [reply projects/plugins/crm/src/automation/class-automation-workflow.php:119]: Fair enough. It's fine for me

File: projects/plugins/crm/src/automation/class-automation-workflow.php
Comment (cleacos) [reply projects/plugins/crm/src/automation/class-automation-workflow.php]: We could create another task for removing the translation from any technical log output.

File: projects/plugins/crm/src/automation/class-automation-boostrap.php
Comment (cleacos) [inline projects/plugins/crm/src/automation/class-automation-boostrap.php]: It seems that `@package` should be used in the file header.

**Review by gogdzl — APPROVED** [review]
It works!

Just as a note, the PR instructions do not work out of the box. It seems that the Contact_Created trigger is listening to the wrong event, I had to change from `jpcrm_automation_contact_created` to `jpcrm_automation_contact_new` in `src/automation/commons/triggers/contacts/class-contact-created.php`.

I don't know if we have any task to change event names, but I thought it is worth mentioning here.

**Pull request merged**
